doc/design/graph.md

```diff
@@ -49,3 +53,18 @@ According to the chain rule of gradient computation, `ConstructBackwardGraph` wo
 For each parameter, like W and b created by `layer.fc`, marked as double circles in above graphs, `ConstructOptimizationGraph` creates an optimization operator to apply its gradient.  Here results in the complete graph:
 
 ![](images/graph_construction_example_all.png)
```

Comment:
Currently we only have **one** kind of directed edge. Which represents both dependency and data flow. It works well but on some case it have problem:
I think we should call A depends on B only if in **every** step running A requires running B.
Not every step running "FC" requires "init" (only the first step needs init),  But currently according to the graph "FC" depends on "W", and "W" depends on "init", thus "FC" depends on "init". I think this kind of "before-after" relationship should be specified by the user, not the graph.

In my opinion we need **two** kinds of directed edges. One for dependency, one for data flow. And maybe for discussion we don't need to draw intermediate var. In the below graph the dotted line is data flow, the full line is dependency. In this representation, there is no cycle in the graph, and "FC" no longer depends on "init" (so we don't need to come up a way to let "init" only run once).

<img width="406" alt="screen shot 2017-09-04 at 4 45 57 pm" src="https://user-images.githubusercontent.com/1724178/30040182-8e4c263a-9190-11e7-93ec-7062cf52fc31.png">

Reply:
Got it. I love this idea and the figure! I agree that there are two kinds of dependencies -- the data dependency and the execution dependency. Currently, we treat them as the same and represent them by the order of operators in array `repeated OpDesc ops` in protobuf message BlockDesc.

I am not sure if it is necessary to explicitly describe these two kinds of dependencies in our protobuf messages.  A reason is that I am not sure what InitAll is -- is it a Var like those returned by operator binding functions, or is it an operator?

Reply:
Sorry, I should have make "init all" more clear. It's an OP that joins / merges all the dependency: will run when all its dependencies are done, it's does nothing itself (only used to join the dependency). Maybe we can call it join or merge.

The reason behind why we need to explicitly describe these two kinds of dependencies is: the PaddlePaddle scheduler only need to schedule OP to run according to the dependency constraint (rather than data flow). For example, in this case, even though "init" writes to var "B", var "B" no longer depends on "init", so when doing optimization, "init" will not be scheduled.

Reply:
Another solution is TF's solution: there is no type Var in the graph. A graph only has OP. A var is represented by a "var OP", which only have output (output the handle for read / write), but no input. And every edge is a tensor rather than var:
<img width="436" alt="screen shot 2017-09-05 at 12 47 42 pm" src="https://user-images.githubusercontent.com/1724178/30080425-6db51fc8-9238-11e7-9816-ef26d23bd58e.png">

Reply:
Record the temporary conclusions from offline discussions:

1. TensorFlow's graph representation embeds variables into operators, and
1. requires users specify input, output, and dependent operators for each operator.

The specification of dependencies looks ugly.  So let's follow our current design of using variables and operators.